ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,405 @@
+        """
+        Maps a given resource to its root for efficient retrieval of the root resource because
+        getting the root resource is likely the most performed operation in `ScriptBuilder`.
+        """
+        resource_id = resource.get_id()
+        if resource_id in self.root_cache:
+            return self.root_cache[resource_id]
+        while len(list(await resource.get_ancestors())) != 0:
+            resource = await resource.get_parent()
+        self.root_cache[resource_id] = resource
+        return resource
+
+    async def _get_variable_from_session(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        return session.get_var_name(resource.get_id())
+
+    async def _var_exists(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        return (
+            resource.get_id() in session.resource_variable_names
+            or resource.get_id() in session.resource_variable_names_queue
+        )
+
+    async def _add_action_to_session_queue(self, resource, action, action_type):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.actions_queue.append(ScriptAction(action_type, action))
+
+    async def _add_variable_to_session_queue(self, resource: Resource, var_name: str) -> None:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.resource_variable_names_queue[resource.get_id()] = var_name
+
+    async def _test_selectable_attributes(self, ancestor, resource, attribute, attribute_value):
+        try:
+            result = await ancestor.get_only_child(
+                r_filter=ResourceFilter(
+                    attribute_filters=[
+                        ResourceAttributeValueFilter(attribute=attribute, value=attribute_value)
+                    ],
+                )
+            )
+        except Exception as e:
+            raise SelectableAttributesError(
+                f"Resource with ID 0x{resource.get_id().hex()} cannot be uniquely identified by attribute "
+                f"{attribute.__name__} (resource has value {attribute_value})."
+            )
+
+    async def _get_selector(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        parent = await resource.get_parent()
+        attribute, attribute_value = await self._get_selectable_attribute(resource)
+        await self._test_selectable_attributes(parent, resource, attribute, attribute_value)
+
+        if isinstance(attribute_value, str) or isinstance(attribute_value, bytes):
+            attribute_value = f'"{attribute_value!s}"'.rstrip()
```

Comment:
```suggestion
            attribute_value = f"{attribute_value!r}"
```
The ["!r"](https://docs.python.org/3.7/tutorial/inputoutput.html) seems like a slightly simpler way to accomplish what we want here, automatically adding quotations if necessary. Though I had to look up "!s" and that's how I found "!r" so maybe you've experimented here a bit more.

In fact I think we can always format the attribute value this way, not just for `str` and `bytes`. Selectable attributes are always indexable attributes which are always primitive types. It has the same result for the other primitives as they are formatted now, but it removes an extra check.

And is there any reason for the `.rstrip()`? At best it seems like it should have no effect because the string is ended with non-whitespace.